cluster.go

```diff
@@ -1135,8 +1124,9 @@ func (c *Cluster) allocate(hash *cid.Cid, repl int) ([]peer.ID, error) {
 
 		// we don't have enough peers to pin
 		if len(candidateAllocs) < needed {
```

Comment:
It's great that you saw that! I saw it too :) You can shortcut the call to allocator.Allocate, but, after calling it, we need to check it anyways.

This is because the `allocator` is a black box (in the future we'd want to support multiple strategies). So even with enough candidates (as in "peers reporting valid metrics"), it could decide against using them (think of an allocator which must allocate to peers in different regions, but all candidates are from the same region). So the output of that call could have less peers than the candidates we give it.

Seeing that, I dropped the `len(candidates) < needed` case because it is a condition that would be caught after calling the allocator anyways.

That said, since you wondered about it, and since the allocator would be a really slow black box, I'll re-add it. Probably makes code more explicit (and marginally more efficient).